File: frontend/src/api/books.ts
import { apiFetch } from './client';

export type BookStatus = 'currently_reading' | 'finished' | 'to_read';

export interface Book {
  id: number;
  open_library_key: string | null;
  title: string;
  author: string | null;
  cover_url: string | null;
  page_count: number | null;
  publish_year: number | null;
  isbn: string | null;
  status: BookStatus;
  year_read: number | null;
  notes: unknown | null;
  created_at: string;
  updated_at: string;
}

export interface SearchResult {
  open_library_key: string | null;
  title: string;
  author: string | null;
  cover_url: string | null;
  page_count: number | null;
  publish_year: number | null;
  isbn: string | null;
}

export interface BookCreate {
  open_library_key?: string | null;
  title: string;
  author?: string | null;
  cover_url?: string | null;
  page_count?: number | null;
  publish_year?: number | null;
  isbn?: string | null;
  status: BookStatus;
}

export function searchBooks(q: string) {
  return apiFetch<SearchResult[]>(`/search?q=${encodeURIComponent(q)}`);
}

export function listBooks(params: { status?: BookStatus; year_read?: number } = {}) {
  const query = new URLSearchParams();
  if (params.status) query.set('status', params.status);
  if (params.year_read) query.set('year_read', String(params.year_read));
  const qs = query.toString();
  return apiFetch<Book[]>(`/books${qs ? `?${qs}` : ''}`);
}

export function getBook(id: number) {
  return apiFetch<Book>(`/books/${id}`);
}

export function createBook(body: BookCreate) {
  return apiFetch<Book>('/books', {
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export function updateBook(id: number, body: Partial<Pick<Book, 'status' | 'year_read' | 'notes'>>) {
  return apiFetch<Book>(`/books/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(body),
  });
}

export function deleteBook(id: number) {
  return apiFetch<void>(`/books/${id}`, { method: 'DELETE' });
}

File: backend/app/routers/books.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.database import get_db
from app.auth import get_current_user
from app.models.user import User
from app.models.book import Book

router = APIRouter(prefix="/books", tags=["books"])


class BookCreate(BaseModel):
    open_library_key: Optional[str] = None
    title: str
    author: Optional[str] = None
    cover_url: Optional[str] = None
    page_count: Optional[int] = None
    publish_year: Optional[int] = None
    isbn: Optional[str] = None
    status: str = "to_read"


class BookUpdate(BaseModel):
    status: Optional[str] = None
    year_read: Optional[int] = None
    notes: Optional[dict] = None


@router.get("")
def list_books(
    status: Optional[str] = Query(None),
    year_read: Optional[int] = Query(None),
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    query = db.query(Book).filter(Book.user_id == user.id)
    if status:
        query = query.filter(Book.status == status)
    if year_read:
        query = query.filter(Book.year_read == year_read)
    return query.order_by(Book.updated_at.desc()).all()


@router.post("", status_code=201)
def create_book(
    body: BookCreate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    book = Book(user_id=user.id, **body.model_dump())
    db.add(book)
    db.commit()
    db.refresh(book)
    return book


@router.get("/{book_id}")
def get_book(
    book_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    book = db.query(Book).filter(Book.id == book_id, Book.user_id == user.id).first()
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")
    return book


@router.patch("/{book_id}")
def update_book(
    book_id: int,
    body: BookUpdate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    book = db.query(Book).filter(Book.id == book_id, Book.user_id == user.id).first()
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")

    update_data = body.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(book, key, value)

    db.commit()
    db.refresh(book)
    return book


@router.delete("/{book_id}", status_code=204)
def delete_book(
    book_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    book = db.query(Book).filter(Book.id == book_id, Book.user_id == user.id).first()
    if not book:
        raise HTTPException(status_code=404, detail="Book not found")
    db.delete(book)
    db.commit()
